src/util/code_observer.h

```diff
@@ -0,0 +1,161 @@
+/// codet in question matches the associated id of the observer.
+///
+/// Example: Replacing all codet(ID_function_call) with codet(ID_skip)
+///
+/// \code
+///
+/// void A::replace_function_calls(codet &code)
+/// {
+///   PRECONDITION(code.get_statement()==ID_function_call);
+///   code = codet(ID_skip);
+/// }
+///
+/// ...
+///
+/// code_observert code_observer;
+/// code_observer.registerObserver(ID_function_call,
+///     &A::replace_function_calls,
+///     this,
+///     std::placeholders::_1);
+/// code_observer.visit_symbols(symbol_table);
+///
+/// ...
+///
+/// \endcode
+///
+/// Note: if the callback only needs to read (and not modify) the codet, then
+/// const_code_observert should be used otherwise use code_observert.
+///
+/// Note': each unique irep_idt can only be associated with one observer.
+
+#ifndef CPROVER_UTIL_CODE_OBSERVER_H
+#define CPROVER_UTIL_CODE_OBSERVER_H
+
+#include <map>
+#include <functional>
+#include <type_traits>
+
+#include "symbol_table.h"
+#include "std_code.h"
+
+/// \brief base class for all code observer implementations
+template<class T> class code_observer_baset
+{
+static_assert(std::is_base_of<codet, T>::value, "T is not derived from codet");
+public:
+  virtual ~code_observer_baset() {}
+
+  /// registers an observer, that is associated with specific codet id.
+  /// \param id: id of the codet to be registered
+  /// \param f: observer callback, that is associated with the given, \p id
+  /// \param args: arguments associated with the given callback, \p f
+  template<class F, class ...Args>
+    void registerObserver(const irep_idt &id, const F &&f, Args&&... args)
+    {
+      auto observer = std::bind(f, args ...);
+      code_observers_map.insert({id, observer});
+    }
+protected:
+  /// notifies (if any) the observer that is registered to receive a callback
+  /// for the given codet, \p code
+  /// \param code: will be passed to the registered observer that
+  /// that has the same associated id
+  void notify(T &code) const
+  {
+    auto it = code_observers_map.find(code.get_statement());
+    if(it != code_observers_map.end())
+      it->second(code);
+  }
+private:
+  using observerst = std::function<void(T&)>;
+  using code_observer_mapt = std::map<const irep_idt, const observerst>;
+  code_observer_mapt code_observers_map;
+};
+
+/// \brief codet observer, uses expression visitor to visit all the codet's
+///   in a given symbol table and appropriately notify registered observers.
+class code_observert final:private expr_visitort,
+  public code_observer_baset<codet>
+{
+public:
+  virtual ~code_observert() {}
+
+  /// iterates through a given symbol table and for every symbol calls it's
+  /// associated expression visitor
+  /// \param symbol_table: symbol table to visit
+  void visit_symbols(symbol_tablet &symbol_table)
+  {
+    forall_symbols(symbol, symbol_table.symbols)
+    {
+      symbolt &w_symbol=symbol_table.get_writeable_ref(symbol->second.name);
```

Comment:
This is a lookup while already iterating over the map! This code should be:
```
Forall_symbols(it, symbol_table.symbols)
{
  it->second.value.visit(*this);
}
```

Reply:
"Forall_symbols" does not exist anymore,

Reply:
Sorry:
```
for(auto &sym_entry : symbol_table.symbols)
{
  sym_entry.second.value.visit(*this);
}
```
(and someone better clean up `symbol_table_baset`, but that's not for this PR)

Reply:
@tautschnig  symbol_table.symbols is constant, so I would need to cast it,  which seems ugly to me.

Reply:
You might want to raise this with @NathanJPhillips or @smowton. I don't recall making or soliciting any of those changes.

Reply:
@tautschnig So three options. 1:) Leave it as it is.  2) Cast, and mess up journaling.  3:) Scrap code_observert, leave const_code_obeservert and leave it up to the actual observer to call get_writable_ref

Reply:
Option 4) fix `symbol_tablet` to provide an interface for iterating over its contained symbols in a non-const setting.

Reply:
agreed, but I think that would be a separate pr no?

Reply:
Possibly, with this one then becoming dependent on it. (Iterating over a container and then doing a lookup of the very same element is just not going to fly.)

Reply:
@tautschnig we will take option (3), thus discarding the non-const iteration in this PR, and wait until a better access to the symbol table is merged, we are quite constrained in terms of time and need to get this merged ASAP.

Reply:
This PR doesn't actually introduce a very essential feature, does it? I'm finding it hard to see the urgency for a half-ready piece of work. The usual approach to visiting various expressions was:

````c++
void do_something_rec(const exprt &expr)
{
  if(expr.id() == ID_code && to_code(expr).get_statement() == ID_foo)
    found_it();
  
  forall_operands(it, expr)
    do_something_rec(*it);
}

void do_something(const symbol_tablet &symbol_table)
{
  for(const auto &sym_entry : symbol_table.symbols)
    do_something_rec(sym_entry.second.value);
}
````

Reply:
The PR introduces a utility that help you avoid such boilerplate every time. That's arguably something desirable (but I agree that it's not absolutely essential (but for the same reason constructs such as functions or classes are not essential either)).

The urgency doesn't stem from the PR itself but from our time constraints in the team ;)

Reply:
I fully understand that there are time constraints. But putting in a half-baked feature that thus only half-solves problems is a very questionable sort of efficiency. Yes, the boilerplate isn't great. But it's not half-baked, it is simple, and not impacted by people fiddling with APIs. And it's not the first of it's kind. All of this boilerplate code ought to be replaced by a decent visitor. _All_ of them, not half of them. I really don't want to see another `depth_iteratort` with its current 2 users.

Reply:
I could do a `symbol_tablet` iterator quickly that had the semantics of a `const_iterator` and so doesn't break sharing/journalling but also included a `get_writeable` method that could be called when an element was encountered that you wanted to write to. Is that what you want? You could build a visitor on top of that, though it would be an unusual one as it would need to implement two callbacks, one that checked whether it wanted to write to the symbol and another that got a writable symbol.

Reply:
@NathanJPhillips  we could do that, and then in the non-const version of the code observer/visitor, when the callback is issued along with a const reference to a codet you will get a reference to the symbol_tabelt iterator which will allows you to call  get_writeable ref.  What do you think? @NathanJPhillips